pkg/loadbalancer/services.go

```diff
@@ -0,0 +1,415 @@
+
+	svc.ServiceParams = params
+	svc.Status = reconciler.StatusPending()
+
+	// Run hooks
+	for _, hook := range s.hooks {
+		hook(txn, &svc)
+	}
+
+	_, _, err = s.svcs.Insert(txn, &svc)
+	return
+}
+
+func (s *Services) DeleteService(txn ServiceWriteTxn, name ServiceName, addr L3n4Addr) (old *Service, err error) {
+	old, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(addr))
+	if !found {
+		return nil, nil
+	}
+	if !old.Name.Equal(name) {
+		return nil, ErrServiceConflict
+	}
+	err = s.deleteService(txn, old)
+	return
+}
+
+func (s *Services) DeleteServicesByName(txn ServiceWriteTxn, name ServiceName, source source.Source) error {
+	svc, _, found := s.svcs.First(txn, ServiceNameIndex.Query(name))
+	if !found {
+		return statedb.ErrObjectNotFound
+	}
+	if svc.Source != source {
+		return ErrServiceSourceMismatch
+	}
+
+	return s.deleteService(txn, svc)
+}
+
+func (s *Services) deleteService(txn ServiceWriteTxn, svc *Service) error {
+	// Release references to the backends
+	iter, _ := s.bes.Get(txn, BackendServiceIndex.Query(svc.Name))
+	for be, _, ok := iter.Next(); ok; be, _, ok = iter.Next() {
+		be, orphan := be.removeRef(svc.Name)
+		if orphan {
+			if _, _, err := s.bes.Delete(txn, be); err != nil {
+				return err
+			}
+		} else {
+			if _, _, err := s.bes.Insert(txn, be); err != nil {
+				return err
+			}
+		}
+	}
+
+	svc.Status = reconciler.StatusPendingDelete()
+	_, _, err := s.svcs.Insert(txn, svc)
+	return err
+}
+
+func (s *Services) DeleteServicesBySource(txn ServiceWriteTxn, source source.Source) error {
+	// Iterating over all as this is a rare operation and it would be costly
+	// to always index by source.
+	iter, _ := s.svcs.All(txn)
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
+		if svc.Source == source {
+			if err := s.deleteService(txn, svc); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Services) UpsertBackends(txn ServiceWriteTxn, serviceName ServiceName, source source.Source, bes ...*Backend) error {
+	if err := s.updateBackends(txn, serviceName, source, bes); err != nil {
+		return err
+	}
+
+	iter, _ := s.svcs.Get(txn, ServiceNameIndex.Query(serviceName))
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
+		svc = svc.Clone()
+		svc.Status = reconciler.StatusPending()
+		svc.BackendRevision = s.bes.Revision(txn)
+		if _, _, err := s.svcs.Insert(txn, svc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func newServiceNameSet(names ...ServiceName) container.ImmSet[ServiceName] {
+	return container.NewImmSetFunc(
+		ServiceName.Compare,
+		names...,
+	)
+}
+
+func (s *Services) updateBackends(txn ServiceWriteTxn, serviceName ServiceName, source source.Source, bes []*Backend) error {
+	for _, bep := range bes {
+		be := ServiceBackend{
+			Backend: *bep,
+			Source:  source,
+		}
+		if old, _, ok := s.bes.First(txn, BackendAddrIndex.Query(be.L3n4Addr)); ok {
+			// UpsertBackends cannot overwrite the backend state or allocated ID.
```

Comment:
Because we might have the same backend added from multiple sources or from multiple EndpointSlices and we want to retain the learned state/health of the backend even when it's being added multiple times. Though now that I look at this again, we do want to propagate the terminating state from K8s. This definitely needs a rethink.